src/som/interpreter/actors/Actor.java

```diff
@@ -190,12 +203,73 @@ protected void appendToMailbox(final EventualMessage msg) {
     mailboxExtension.append(msg);
   }
 
-  protected static void handleBreakPoints(final EventualMessage msg, final WebDebugger dbg) {
-    if (VmSettings.TRUFFLE_DEBUGGER_ENABLED && msg.isMessageReceiverBreakpointSet()) {
+  protected static void handleBreakpointsAndStepping(final EventualMessage msg, final WebDebugger dbg, final Actor actor) {
+    if (!VmSettings.TRUFFLE_DEBUGGER_ENABLED) {
+      return;
+    }
+
+    synchronized (actor) {
```

Comment:
If you look at the code (the old code you referenced also needs to ensure that, a bit simpler, but not up-to-date) https://github.com/smarr/SOMns/blob/master/src/som/interpreter/actors/Actor.java#L292-L315, you will see that the state of the actor object is updated. The mailbox is exchanged, and the status of the actor is set to indicate that it is currently executing.

Since the mailbox is where all other threads/actors converge to sequentialize, it needs to be protected (as well as the state indicating whether an actor is executing or not).

It is unclear to my why you need to synchronize on the actor here, however.
Conceptually, there can only be a single thread executing an actor at a time.
So, all stepping operations that work on an actor are done by a single thread.
Consequently, I don't see a need why it needs to be synchronized.